File: split/copynodelist.pas
function copynodelist(p: halfword): halfword;
var
  h: halfword;
  q: halfword;
  r: halfword;
  words: 0..5;
begin
  h := getavail;
  q := h;
  while p <> 0 do
  begin
    {205:}
words := 1;
    if (p >= himemmin) then
      r := getavail
    else
      {206:}
case mem[p].hh.b0 of
        0, 1, 13:
        begin
          r := getnode(7);
          mem[r + 6] := mem[p + 6];
          mem[r + 5] := mem[p + 5];
          mem[r + 5].hh.rh := copynodelist(mem[p + 5].hh.rh);
          words := 5;
        end;
        2:
        begin
          r := getnode(4);
          words := 4;
        end;
        3:
        begin
          r := getnode(5);
          mem[r + 4] := mem[p + 4];
          mem[mem[p + 4].hh.rh].hh.rh := mem[mem[p + 4].hh.rh].hh.rh + 1;
          mem[r + 4].hh.lh := copynodelist(mem[p + 4].hh.lh);
          words := 4;
        end;
        8:{1357:}
case mem[p].hh.b1 of
            0:
            begin
              r := getnode(3);
              words := 3;
            end;
            1, 3:
            begin
              r := getnode(2);
              mem[mem[p + 1].hh.rh].hh.lh := mem[mem[p + 1].hh.rh].hh.lh + 1;
              words := 2;
            end;
            2, 4:
            begin
              r := getnode(2);
              words := 2;
            end;
            others: confusion(1293)
          end{:1357}
;
        10:
        begin
          r := getnode(2);
          mem[mem[p + 1].hh.lh].hh.rh := mem[mem[p + 1].hh.lh].hh.rh + 1;
          mem[r + 1].hh.lh := mem[p + 1].hh.lh;
          mem[r + 1].hh.rh := copynodelist(mem[p + 1].hh.rh);
        end;
        11, 9, 12:
        begin
          r := getnode(2);
          words := 2;
        end;
        6:
        begin
          r := getnode(2);
          mem[r + 1] := mem[p + 1];
          mem[r + 1].hh.rh := copynodelist(mem[p + 1].hh.rh);
        end;
        7:
        begin
          r := getnode(2);
          mem[r + 1].hh.lh := copynodelist(mem[p + 1].hh.lh);
          mem[r + 1].hh.rh := copynodelist(mem[p + 1].hh.rh);
        end;
        4:
        begin
          r := getnode(2);
          mem[mem[p + 1].int].hh.lh := mem[mem[p + 1].int].hh.lh + 1;
          words := 2;
        end;
        5:
        begin
          r := getnode(2);
          mem[r + 1].int := copynodelist(mem[p + 1].int);
        end;
        others: confusion(354)
      end{:206}
;
    while words > 0 do
    begin
      words := words - 1;
      mem[r + words] := mem[p + words];
    end{:205}
;
    mem[q].hh.rh := r;
    q := r;
    p := mem[p].hh.rh;
  end;
  mem[q].hh.rh := 0;
  q := mem[h].hh.rh;
  begin
    mem[h].hh.rh := avail;
    avail := h;{dynused:=dynused-1;}

  end;
  copynodelist := q;
end;

File: cpp/copynodelist.cpp
#include "copynodelist.h"
#include "getavail.h"
#include "getnode.h"
#include "confusion.h"

static void add_token_ref(halfword p)
{
	token_ref_count(p)++;
}

halfword copynodelist(halfword p)
{
	auto h = getavail();
	auto q = h;
	while (p)
	{
		int words = 1;
		halfword r;
		if (p >= himemmin)
			r = getavail();
		else
			switch (type(p))
			{
				case hlist_node:
				case vlist_node:
				case unset_node:
					r = getnode(box_node_size);
					mem[r+6] = mem[p+6];
					mem[r+5] = mem[p+5];
					list_ptr(r) = copynodelist(list_ptr(p));
					words = 5;
					break;
				case rule_node:
					r = getnode(rule_node_size);
					words = rule_node_size;
					break;
				case ins_node:
					r = getnode(ins_node_size);
					mem[r+4] = mem[p+4];
					add_glue_ref(split_top_ptr(p));
					ins_ptr(r) = copynodelist(ins_ptr(p));
					words = ins_node_size-1;
					break;
				case whatsit_node:
					switch (subtype(p))
					{
						case open_node:
							r = getnode(open_node_size);
							words = open_node_size;
							break;
						case write_node:
						case special_node:
							r = getnode(write_node_size);
							add_token_ref(write_tokens(p));
							words = write_node_size;
							break;
						case close_node:
						case language_node:
							r = getnode(small_node_size);
							words = small_node_size;
							break;
						default:
							confusion("ext2");
					}
					break;
				case glue_node:
					r = getnode(small_node_size);
					add_glue_ref(glue_ptr(p));
					glue_ptr(r) = glue_ptr(p);
					leader_ptr(r) = copynodelist(leader_ptr(p));
					break;
				case kern_node:
				case math_node:
				case penalty_node:
					r = getnode(small_node_size);
					words = 2;
					break;
				case ligature_node:
					r = getnode(small_node_size);
					mem[lig_char(r)] = mem[lig_char(p)];
					lig_ptr(r) = copynodelist(lig_ptr(p));
					break;
				case disc_node:
					r = getnode(small_node_size);
					pre_break(r) = copynodelist(pre_break(p));
					post_break(r) = copynodelist(post_break(p));
					break;
				case mark_node:
					r = getnode(small_node_size);
					info(mark_ptr(p))++;
					words = small_node_size;
					break;
				case adjust_node:
					r = getnode(small_node_size);
					adjust_ptr(r) = copynodelist(adjust_ptr(p));
					break;
				default: 
					confusion("copying");
			}
		;
		for (;words > 0; words--)
			mem[r+words] = mem[p+words];
		link(q) = r;
		q = r;
		p = link(p);
	}
	link(q) = 0;
	q = link(h);
	link(h) = avail;
	avail = h;
	return q;
}
